src/function/maybe_changed_after.rs

```diff
@@ -170,88 +251,105 @@ where
-                            // so even if we mark them as valid here, the function will re-execute
-                            // and overwrite the contents.
-                            dependency_index
-                                .mark_validated_output(db.as_dyn_database(), database_key_index);
+                            QueryEdge::Output(dependency_index) => {
+                                // Subtle: Mark outputs as validated now, even though we may
+                                // later find an input that requires us to re-execute the function.
+                                // Even if it re-execute, the function will wind up writing the same value,
+                                // since all prior inputs were green. It's important to do this during
+                                // this loop, because it's possible that one of our input queries will
+                                // re-execute and may read one of our earlier outputs
+                                // (e.g., in a scenario where we do something like
+                                // `e = Entity::new(..); query(e);` and `query` reads a field of `e`).
+                                //
+                                // NB. Accumulators are also outputs, but the above logic doesn't
+                                // quite apply to them. Since multiple values are pushed, the first value
+                                // may be unchanged, but later values could be different.
+                                // In that case, however, the data accumulated
+                                // by this function cannot be read until this function is marked green,
+                                // so even if we mark them as valid here, the function will re-execute
+                                // and overwrite the contents.
+                                //
+                                // TODO not if we found a cycle head other than ourself?
+                                dependency_index.mark_validated_output(
+                                    db.as_dyn_database(),
+                                    database_key_index,
+                                );
+                            }
                         }
                     }
+                    inputs
                 }
-                inputs
-            }
-        };
+            };
 
-        old_memo.mark_as_verified(
-            db.as_dyn_database(),
-            zalsa.current_revision(),
-            database_key_index,
-            inputs,
-        );
-        true
+            let in_heads = cycle_heads.remove(&database_key_index);
+
+            if cycle_heads.is_empty() {
+                old_memo.mark_as_verified(
+                    db.as_dyn_database(),
+                    zalsa.current_revision(),
+                    database_key_index,
+                    inputs,
+                );
+            }
+            if in_heads {
```

Comment:
Would you mind adding a comment explaining why we clear the cycle heads if the query itself is in the cycle heads?

Reply:
The surface reason is pretty simple: we are going to do a fresh traversal of this memo's dependencies, so we need to start with a clean slate. I previously created a fresh hashset each loop, and in an earlier review you wisely suggested to just reuse one hashset by clearing it.

But thanks for suggesting I document this! Because trying to write this comment led me down a rabbit hole of finding and fixing several bugs, and commenting all of this `deep_verify_memo` logic much more thoroughly.